Support Rfc3279 signature format for DSA and EcDSA

krwq opened the pull request:
Fixes: #31548

Please see the issue for more details.

krwq committed:
Support DSASignatureFormat

bartonjs commented on src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EcDsa.cs:
```diff
@@ -10,12 +10,12 @@ internal static partial class Interop
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EcDsaSign")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool EcDsaSign(ref byte dgst, int dlen, ref byte sig, [In, Out] ref int siglen, SafeEcKeyHandle ecKey);
+        private static extern bool EcDsaSign(ref byte dgst, int dlen, ref byte sig, out int siglen, SafeEcKeyHandle ecKey);
```
FWIW, I think this accidentally worked because you used the same variable as when it said ref, and the compiler feels no real reason to clear a variable being passed to an out.

krwq replied:
@bartonjs, yes value seem to be used for conversion but then I think ECDSA_sign would ignore it anyway - perhaps I'm missing something https://www.openssl.org/docs/man1.1.0/man3/ECDSA_sign.html but only mentions out usage. If that's the case I'll remove unnecessary line in the shim

bartonjs replied:
Hm, it does appear that they only write to it in 1.0.2 and 3.0, presumably they didn't change that for 1.1.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Helpers.cs:
```diff
@@ -84,5 +84,39 @@ internal static void ConvertIntToByteArray(uint value, byte[] dest)
+
+        public static bool TryGetSizeInBytes(this HashAlgorithmName hashName, out int size)
+        {
+            if (hashName.Name == HashAlgorithmName.SHA256.Name)
```
Why not `if (hashName == HashAlgorithmName.SHA256)`?

krwq replied:
I think I initially wanted to first ref equals and after that string comparison but then simplified it since it's cheap comparing to operation itself, will change

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -79,27 +79,66 @@ public byte[] SignData(byte[] data, HashAlgorithmName hashAlgorithm)
+            {
+                throw new ArgumentNullException(nameof(data));
+            }
+
```
You should also check that signatureFormat is one of the known inputs.  Especially since the enum and this type are in the same assembly, they can't skew.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -79,27 +79,66 @@ public byte[] SignData(byte[] data, HashAlgorithmName hashAlgorithm)
+            return SignDataCore(new ReadOnlySpan<byte>(data, offset, count), hashAlgorithm, signatureFormat);
+        }
+
+        protected virtual byte[] SignDataCore(ReadOnlySpan<byte> data, HashAlgorithmName hashAlgorithm, DSASignatureFormat signatureFormat)
```
In a manner left as an exercise to the reader: I'm concerned that the current changes won't work for existing out-of-box implementations that override the current virtuals.  Make sure there's a test that calling the new method specifying the old format does sensible things with netstandard2.1-based derived types.

krwq replied:
I believe they should work since I always end up calling abstract method as a fallback if it's not overriden (there is even an existing test which tests some of that scenarios)

bartonjs replied:
> there is even an existing test which tests some of that scenarios

Oh, good.  Yeah, I hadn't made it through the call chain yet to see that there was that fallback and if it was going to be complicated wanted to get the feedback out there before I had to run.

bartonjs reviewed (commented):
Partial review, will resume later.

krwq commented on src/libraries/Common/src/System/Security/Cryptography/DSAOpenSsl.cs:
```diff
@@ -234,45 +234,60 @@ public override byte[] CreateSignature(byte[] rgbHash)
+                    try
                     {
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                        bool success = Interop.Crypto.DsaSign(key, hash, new Span<byte>(signature, 0, signatureSize), out signatureSize);
```
Per my understanding of OpenSSL docs it's actually out and the docs says `The length of the signature is places in *siglen. sigret must point to DSA_size(dsa) bytes of memory.` - that's why I added the hard check even though sometimes signature might be tiny bit smaller

krwq commented on src/libraries/Common/src/System/Security/Cryptography/DSAOpenSsl.cs:
```diff
@@ -234,45 +234,60 @@ public override byte[] CreateSignature(byte[] rgbHash)
+
+                    if (!success)
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
```
Yes, I missed the hardware keys scenarios, I though this was just equivalent to property getter

bartonjs replied:
Calling GetKey also causes a new software key to be generated, if it wasn't already.  DSA-3072 sometimes takes a while, and it'd be frustrating to get the ArgumentException after waiting on keygen :)

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -118,10 +157,18 @@ public virtual bool VerifyData(byte[] data, int offset, int count, byte[] signat
+            if (offset < 0 || offset > data.Length) { throw new ArgumentOutOfRangeException(nameof(offset)); }
+            if (count < 0 || count > data.Length - offset) { throw new ArgumentOutOfRangeException(nameof(count)); }
+            if (signature == null) { throw new ArgumentNullException(nameof(signature)); }
+
```
And validate signatureFormat (applies to multiple methods)

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -118,10 +157,18 @@ public virtual bool VerifyData(byte[] data, int offset, int count, byte[] signat
             if (offset < 0 || offset > data.Length) { throw new ArgumentOutOfRangeException(nameof(offset)); }
             if (count < 0 || count > data.Length - offset) { throw new ArgumentOutOfRangeException(nameof(count)); }
             if (signature == null) { throw new ArgumentNullException(nameof(signature)); }
-            if (string.IsNullOrEmpty(hashAlgorithm.Name)) { throw HashAlgorithmNameNullOrEmpty(); }
```
Why was this check removed?

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -79,27 +79,66 @@ public byte[] SignData(byte[] data, HashAlgorithmName hashAlgorithm)
+            Debug.Assert(size <= 256, "GetMaxSignatureSize returned more than expected");
+            Span<byte> signature = stackalloc byte[size];
+            bool result = TrySignDataCore(data, signature, hashAlgorithm, signatureFormat, out int bytesWritten);
+            Debug.Assert(result, "GetMaxSignatureSize returned insufficient size");
```
I'd be more comfortable with

```C#
if (!TrySignDataCore(...))
{
    Debug.Fail("GetMaxSignatureSize returned insufficient size");
    throw new CryptographicException();
}

return signature.Slice(0, bytesWritten).ToArray();
```

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -130,24 +177,43 @@ public virtual bool VerifyData(Stream data, byte[] signature, HashAlgorithmName
+
+        protected virtual bool TryCreateSignatureCore(ReadOnlySpan<byte> hash, Span<byte> destination, DSASignatureFormat signatureFormat, out int bytesWritten)
+        {
+            // This method is expected to be overriden with better implementation
```
When we update the docs for this class will we recommend falling back to this implementation for an unknown signature format, since we'd theoretically handle it in ConvertIeeeP1363Signature for them?

Alternatively, do we want to add something so a derived type can report whether or not it directly handles a format, so that we understand when to apply an 1363-to-destination-scheme strategy?

bartonjs replied:
(Maybe it's moot since there're only the two, but I like to think ahead)

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -195,44 +262,97 @@ public virtual bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination,
+        protected virtual bool VerifyDataCore(Stream data, ReadOnlySpan<byte> signature, HashAlgorithmName hashAlgorithm, DSASignatureFormat signatureFormat)
         {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
```
There shouldn't be any validation in a Core method, it should all be in the public entrypoint.  That way overrides don't need to reimplement it.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -195,44 +262,97 @@ public virtual bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination,
+            byte[] sig = this.ConvertSignatureToIeeeP1363(signatureFormat, signature);
+
+            if (sig == null)
+                return false;
```
Why would we get back null? (Add a comment)

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -419,5 +539,22 @@ private AsnWriter WriteSubjectPublicKeyInfo()
+                    // Add 15 for extra ASN.1 headers
+                    return qLength * 2 + 15;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(signatureFormat));
```
A) I think we want to add a protected virtual property or method to get the size of Q (much cheaper than ExportParameters), which can default to calling ExportParameters.

B) Prefer throwing the argument before calling that member, since ExportParameters will generate a key if it hasn't already happened.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSASignatureFormat.cs:
```diff
@@ -0,0 +1,12 @@
+    public enum DSASignatureFormat
+    {
+        IeeeP1363FixedFieldConcatenation,
+        Rfc3279DerSequence
```
Please add the trailing comma.

Also, please go ahead and write xmldoc for these.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsa.cs:
```diff
@@ -86,8 +90,83 @@ public virtual byte[] SignData(byte[] data, int offset, int count, HashAlgorithm
+        {
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+            {
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
```
Same general feedback: There should rarely be argument validation in a Core method (only validation that applies to that specific derived implementation).  Common validation should be in the non-Core method.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsa.cs:
```diff
@@ -107,6 +186,28 @@ public virtual bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination,
+            }
+
+            if (TryHashData(data, destination, hashAlgorithm, out int hashLength) &&
+                TrySignHashCore(destination.Slice(0, hashLength), destination, signatureFormat, out bytesWritten))
```
I don't know that we want to depend on the implementation of TrySignHashCore to support signing in-place (that it can, without fail, write a signature over top of the hash value memory).

Please hash into a temporary location (stackalloc/rent).  This also applies to DSA if I missed it there.

krwq replied:
I was originally planning to do that but I noticed we already did that before and followed the pattern but will fix

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -419,5 +539,22 @@ private AsnWriter WriteSubjectPublicKeyInfo()
+                    return qLength * 2;
+                case DSASignatureFormat.Rfc3279DerSequence:
+                    // Add 15 for extra ASN.1 headers
+                    return qLength * 2 + 15;
```
If we're using this as the lower bound of what someone has to pass in (at least for the OpenSsl implementation) then maybe it needs to be a more accurate max.

// 30 7E 02 3E 00 [qLength] 02 3E 00 [qlength]
* qLength <= 0x3D (62) ? qLength * 2 + 8;
// 30 81 FE 02 7C 00 [qLength] 02 7C 00 [qLength]
* qLength <= 0x7B (124) ? qLength * 2 + 9;
// 30 82 01 00 02 7F 00 [qLength] 02 7F 00 [qLength]
* qLength <= 0x7E (126) ? qLength * 2 + 10
// 30 82 02 02 02 81 FF 00 [qLength] 02 81 FF 00 [qLength]
* qLength <= 0xFE (254) ? qLength * 2 + 12
// 30 82 FF FE 02 82 7F FD 00 [qLength] 02 82 7F FD 00 [qLength]
* qLength <= 0x7FFC (32764) ? qLength * 2 + 14
// 30 83 02 00 02 02 82 FF FF 00 [qLength] 02 82 FF FF 00 [qLength]
* qLength <= 0xFFFE (65534) ? qLength * 2 + 15
// 30 83 FF FF FE 02 83 7F FF FD 00 [qLength] 02 83 7F FF FD 00 [qLength]
* qLength <= 0x7F_FFFC (8_388_604) ? qLength * 2 + 17
// 30 84 02 00 00 02 02 83 FF FF FF 00 [qLength] 02 83 FF FF FF 00 [qLength]
* qLength <= 0xFF_FFFE (16_777_214) ? qLength * 2 + 18
// 30 84 FF FF FF FE 02 84 7F FF FF FD 00 [qLength] 02 84 7F FF FF FD 00 [qLength]
* qLength <= 0x7FFF_FFFC (2_147_483_644) ? qLength + 20

Beyond here we need the CLR to support bigger arrays.  Probably could have stopped a bit ago :)

bartonjs replied:
Also, FWIW, the biggest current legal size of qLength is 32.  (FIPS 186-4 says the biggest legal DSA key size has N=256, and N is the length of q, in bits).

krwq committed:
Do not hash and sign on the same buffer, improve diagnostics

krwq commented:
Marking as NO REVIEW until I figure out all the failures and fix build issues on non-Windows

krwq committed:
Build on Linux

krwq committed:
ignore new overloads should be overriden in csp tests

krwq committed:
provide impl of CSP on linux

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/tests/DsaEcDsaSignatureFormatTests.cs:
```diff
@@ -0,0 +1,1166 @@
+                byte[] sig = Sign(test, signatureFormat);
+
+                SetData(IsSigner(), offset, count);
+                Assert.True(Verify(test, sig, signatureFormat));
```
This test failed once... but we don't know which walk through a foreach or why.  Consider upgrading all of the Assert.True (or Assert.False) calls to use the message parameter to report what they're testing.  Remember that the message is displayed as

```
Message
    Expected: True
    Actual: False
```

So this one would be

```C#
Assert.True(
    Verify(test, sig, signatureFormat),
    $"Signature verifies for Slice({offset}, {count}) when IsSigner is {IsSigner()}");
```

carlossanlop commented on src/libraries/System.Security.Cryptography.Algorithms/ref/System.Security.Cryptography.Algorithms.cs:
```diff
@@ -126,6 +126,8 @@ public abstract partial class DSA : System.Security.Cryptography.AsymmetricAlgor
         public static System.Security.Cryptography.DSA Create(System.Security.Cryptography.DSAParameters parameters) { throw null; }
         public static new System.Security.Cryptography.DSA Create(string algName) { throw null; }
         public abstract byte[] CreateSignature(byte[] rgbHash);
+        public byte[] CreateSignature(byte[] rgbHash, System.Security.Cryptography.DSASignatureFormat signatureFormat) { throw null; }
```
@krwq this PR is adding new public APIs. Please make sure to document them with triple slash in their implementation so we can port them in the future to dotnet-api-docs.

cc @bartonjs in case you'll follow up with this PR.

bartonjs committed:
Merge remote-tracking branch 'dotnet/master' into ecdsa-signature-format

bartonjs committed:
Make the allConfigurations build work again

bartonjs committed:
Improve test determinism

bartonjs committed:
Add missing format conversion to TrySignHashCore

bartonjs committed:
Refactor DsaEcDsaSignatureFormatTests and move to Common

This accounts for:

* macOS not supporting DSA key generation
* macOS not supporting FIPS 186-3 DSA
* DSACryptoServiceProvider not supporting FIPS 186-3 DSA
* Testing the signature format tests on the concrete types, as well as the opaque ones.

bartonjs committed:
Fix explicit key parameter set to one with a private key

bartonjs committed:
Enable expected-size writes for the OpenSsl types

bartonjs committed:
Start cleanup

stephentoub commented:
@bartonjs, looks like you've taken this PR over? Is it still "no merge"?

bartonjs committed:
Fix infinite recursion in DSAOpenSsl

bartonjs committed:
Restore some overrides that got lost thus far.

bartonjs committed:
Use a stack-span more in DSAOpenSsl.TryCreateSignature(Core)

bartonjs committed:
Restore lost overrides in DSACng

bartonjs committed:
Restore lost overrides in ECDsaCng

bartonjs committed:
Improve stackalloc usage, centralize error handling

bartonjs committed:
Restore lost overrides in ECDsaOpenSsl

bartonjs committed:
Doing math right is important.

bartonjs committed:
Tighten in GetMaxDerSignatureSize values

bartonjs committed:
Normalize ECDsa argument validation, other cleanup

bartonjs committed:
CryptoNative_EcDsaSign should treat siglen as out, not ref.

bartonjs committed:
Remove some more temporary arrays

bartonjs committed:
Don't override new members in Unix DSACSP

bartonjs committed:
Housekeeping

bartonjs committed:
Don't depend on zero-init destinations

bartonjs committed:
Merge remote-tracking branch 'dotnet/master' into ecdsa-signature-format

bartonjs committed:
React to nullability changes

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -76,30 +82,108 @@ protected virtual byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw HashAlgorithmNameNullOrEmpty();
+
+            return SignDataCore(data, hashAlgorithm, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);
```
This changes the virtual dispatch order (used to be `SignData(byte[], HashAlgorithmName)` => `virtual SignData(byte[], int, int, HashAlgorithmName)` => `virtual HashData(byte[], int, int)` + `abstract CreateSignature(byte[])` now it's `SignData(byte[], HashAlgorithmName)` => `virtual SignDataCore(ReadOnlySpan<byte>, HashAlgorithmName, DSASignatureFormat)` => `virtual TrySignData(ReadOnlySpan<byte>, Span<byte>, HashAlgorithmName, DSASignatureFormat, out int)` => `virtual TryHashData(...)` (=> `virtual HashData(...)`) + `virtual TryCreateSignature(...)` (=> `abstract CreateSignature(byte[])`. -- The important part of the change is we no longer ever call `virtual SignData(byte[], int, int, HashAlgorithmName)`.

For .NET Framework we'd say no to a change like that.  I'm not sure how we feel in .NET Core.

bartonjs committed:
Merge remote-tracking branch 'dotnet/master' into ecdsa-signature-format

bartonjs committed:
Re-normalize ref.cs

bartonjs committed:
Fix Cng.Tests build for net472

bartonjs committed:
Add API docs for DSASignatureFormat and DSA

bartonjs committed:
Remove inaccurate exception docs

bartonjs committed:
Add units to GetMaxSignatureSize docs

bartonjs committed:
Add docs to new ECDsa members

bartonjs committed:
Restore the order of virtual method calls in DSA

The existing methods should all be unchanged, and the new methods written
in terms of the existing ones.

bartonjs committed:
Add back the TryCreateSignature and VerifySignature overrides

bartonjs committed:
Restore ECDsa virtual method call ordering

GrabYourPitchforks commented on src/libraries/Common/src/System/Security/Cryptography/ECDsaCng.SignVerify.cs:
```diff
@@ -48,12 +46,57 @@ public override byte[] SignHash(byte[] hash)
+
+            return AsymmetricAlgorithmHelpers.TryConvertIeee1363ToDer(
+                destination.Slice(0, bytesWritten),
+                destination,
```
Was it intended that these buffers overlap? Is the underlying behavior well-defined for this scenario?

bartonjs replied:
The calculation uses an intermediate already, so it's safe to overlap (it's intended for this usage)

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -194,45 +541,371 @@ public virtual bool TrySignData(ReadOnlySpan<byte> data, Span<byte> destination,
+            finally
+            {
+                Array.Clear(array, 0, data.Length);
+                ArrayPool<byte>.Shared.Return(array);
```
It's fine to have the `Array.Clear` inside the _finally_ block, but general guidance is that we shouldn't return the array to the pool in exceptional cases since we should assume the buffer is in an indeterminate state. Let the GC reclaim it.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsa.cs:
```diff
@@ -90,21 +99,355 @@ public virtual byte[] SignData(byte[] data, int offset, int count, HashAlgorithm
+                if (rented != null)
+                {
+                    CryptographicOperations.ZeroMemory(rented.AsSpan(0, bytesWritten));
+                    ArrayPool<byte>.Shared.Return(rented);
```
Same comment about `ArrayPool` and _finally_ blocks here. (This comment applies elsewhere throughout these files.)

GrabYourPitchforks reviewed (approved):
Some questions about edge cases and error handling, but otherwise seems solid. :)

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs:
```diff
@@ -20,45 +18,170 @@ internal static partial class AsymmetricAlgorithmHelpers
+
+            static int GetDerLengthLength(int payloadLength)
+            {
+                Debug.Assert(payloadLength >= 0);
```
The lengths always originate with either ecdsa.KeySize or dsa.ExportParameters(false).Q.Length; both of which are class-controlled instead of payload-controlled.  So I think it's OK.

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs:
```diff
@@ -69,18 +192,75 @@ private static void CopySignatureField(ReadOnlySpan<byte> signatureField, Span<b
-                Debug.Assert(signatureField[1] > 0x7F, "signatureField[1] > 0x7F");
+                if (signatureField.Length != response.Length + 1 ||
+                    signatureField[0] != 0 ||
+                    signatureField[1] <= 0x7F)
```
Response is always field-size wide; so an empty response would mean a zero-bit key.

SignatureField can't represent zero, for algorithm-conforming types, because neither r nor s are allowed to be zero in either DSA or ECDSA.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/DSACng.SignVerify.cs:
```diff
@@ -63,14 +97,41 @@ public override bool VerifySignature(byte[] rgbHash, byte[] rgbSignature)
+                if (signatureFormat == DSASignatureFormat.Rfc3279DerSequence)
+                {
+                    // source.Length is the field size, in bytes, so just convert to bits.
+                    int fieldSizeBits = source.Length * 8;
```
Source is bounded by 32 bytes, but is always whatever the size of Q is.  (It's either exactly right, left-padded with zeros, or truncated to the left Q-size bytes)

bartonjs committed:
Apply feedback
